File: day_03.rb
def traverse(col_movement, row_movement, forest)
  row_index = col_index = tree_count = 0
  while row_index < forest.length - 1
    row_index += row_movement
    row = forest[row_index]
    col_index = (col_index + col_movement) % row.length
    tree_count += 1 if row[col_index] == '#'
  end
  tree_count
end

text = File.open('day_03_input.txt').read
rows = []
text.each_line do |row|
  cols = []
  row.strip.each_char do |col|
    cols.push(col)
  end
  rows.push(cols)
end

part_1 = traverse(3, 1, rows)
puts "Part 1: #{part_1}"
puts "Part 2: #{traverse(1, 1, rows) *
    traverse(3, 1, rows) *
    traverse(5, 1, rows) *
    traverse(7, 1, rows) *
    traverse(1, 2, rows)}"

File: day_03.py
def traverse(col_movement, row_movement):
    row_index = 0
    col_index = 0
    tree_count = 0
    while row_index < len(rows) - 1:
        row_index += row_movement
        row = rows[row_index]
        col_index = (col_index + col_movement) % len(row)
        col_char = row[col_index]
        if col_char == '#':
            tree_count += 1
    return tree_count


with open('day_03_input.txt', 'r') as f:
    raw = f.readlines()

rows = []
for line in raw:
    cols = []
    for c in line.strip():
        cols.append(c)
    rows.append(cols)

tree_count_1_1 = traverse(1, 1)
tree_count_3_1 = traverse(3, 1)  # Part 1 solution
tree_count_5_1 = traverse(5, 1)
tree_count_7_1 = traverse(7, 1)
tree_count_1_2 = traverse(1, 2)

part_2_solution = tree_count_1_1 * tree_count_3_1 * tree_count_5_1 * tree_count_7_1 * tree_count_1_2

print("Part 1: Tree count = {}".format(tree_count_3_1))
print("Part 2: Solution = {}".format(part_2_solution))
